File: go/internal/model/actions/actions_toolcall.go
// Package actions provides action parsing and observation formatting.
// Ported from src/minisweagent/models/utils/actions_toolcall.py
package actions

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/tonythethompson/mini-swe-agent/go/internal/exceptions"
	"github.com/tonythethompson/mini-swe-agent/go/internal/utils/jinja"
)

// BashTool is the OpenAI tool definition for the bash command.
var BashTool = map[string]any{
	"type": "function",
	"function": map[string]any{
		"name":        "bash",
		"description": "Execute a bash command",
		"parameters": map[string]any{
			"type": "object",
			"properties": map[string]any{
				"command": map[string]any{
					"type":        "string",
					"description": "The bash command to execute",
				},
			},
			"required": []string{"command"},
		},
	},
}

// ParseToolcallActions parses tool calls from the response.
// Returns FormatError if unknown tool or invalid args.
func ParseToolcallActions(toolCalls []exceptions.ToolCall, formatErrorTemplate string, templateKwargs map[string]any) ([]exceptions.Action, error) {
	if len(toolCalls) == 0 {
		content := jinja.Render(formatErrorTemplate, mergeMaps(map[string]any{
			"error":           "No tool calls found in the response. Every response MUST include at least one tool call.",
			"actions":         []any{},
			"has_tool_calls":  false,
		}, templateKwargs))
		return nil, exceptions.NewFormatError(exceptions.Message{
			Role:    "user",
			Content: content,
			Extra:   map[string]any{"interrupt_type": "FormatError"},
		})
	}
	var actions []exceptions.Action
	for _, tc := range toolCalls {
		var errMsg string
		var args map[string]any
		if err := json.Unmarshal([]byte(tc.Function.Arguments), &args); err != nil {
			errMsg = fmt.Sprintf("Error parsing tool call arguments: %v.", err)
		}
		if tc.Function.Name != "bash" {
			errMsg += fmt.Sprintf("Unknown tool '%s'.", tc.Function.Name)
		}
		if args == nil || args["command"] == nil {
			errMsg += "Missing 'command' argument in bash tool call."
		}
		if errMsg != "" {
			content := jinja.Render(formatErrorTemplate, mergeMaps(map[string]any{
				"actions":        []any{},
				"error":           errMsg,
				"has_tool_calls":  true,
			}, templateKwargs))
			return nil, exceptions.NewFormatError(exceptions.Message{
				Role:    "user",
				Content: content,
				Extra:   map[string]any{"interrupt_type": "FormatError"},
			})
		}
		actions = append(actions, exceptions.Action{
			Command:    args["command"].(string),
			ToolCallID: tc.ID,
		})
	}
	return actions, nil
}

// FormatToolcallObservationMessages formats execution outputs into tool result messages.
func FormatToolcallObservationMessages(actions []exceptions.Action, outputs []exceptions.EnvOutput, observationTemplate string, templateVars map[string]any, multimodalRegex string) []exceptions.Message {
	notExecuted := exceptions.EnvOutput{Output: "", ReturnCode: -1, ExceptionInfo: "action was not executed"}
	paddedOutputs := make([]exceptions.EnvOutput, len(actions))
	for i := range paddedOutputs {
		if i < len(outputs) {
			paddedOutputs[i] = outputs[i]
		} else {
			paddedOutputs[i] = notExecuted
		}
	}
	var results []exceptions.Message
	for i, action := range actions {
		output := paddedOutputs[i]
		content := jinja.Render(observationTemplate, mergeMaps(map[string]any{"output": output}, templateVars))
		msg := exceptions.Message{
			Content: content,
			Extra: map[string]any{
				"raw_output":      output.Output,
				"returncode":      output.ReturnCode,
				"timestamp":       float64(time.Now().UnixNano()) / 1e9,
				"exception_info":  output.ExceptionInfo,
			},
		}
		for k, v := range output.Extra {
			msg.Extra[k] = v
		}
		if action.ToolCallID != "" {
			msg.ToolCallID = action.ToolCallID
			msg.Role = "tool"
		} else {
			msg.Role = "user"
		}
		results = append(results, msg)
	}
	return results
}

func mergeMaps(base, overlay map[string]any) map[string]any {
	result := make(map[string]any, len(base)+len(overlay))
	for k, v := range base {
		result[k] = v
	}
	for k, v := range overlay {
		result[k] = v
	}
	return result
}

File: ts/src/models/utils/actions_toolcall.ts
/** Parse actions & format observations with tool calls.
 * Ported from src/minisweagent/models/utils/actions_toolcall.py */
import nunjucks from "nunjucks";
import { FormatError, type Action, type EnvOutput, type Message, type ToolCall } from "../../exceptions.js";
import { expandMultimodalContent } from "./multimodal.js";

export const BASH_TOOL = {
  type: "function",
  function: {
    name: "bash",
    description: "Execute a bash command",
    parameters: {
      type: "object",
      properties: {
        command: { type: "string", description: "The bash command to execute" },
      },
      required: ["command"],
    },
  },
};

/** Parse tool calls from the response. Raises FormatError if unknown tool or invalid args. */
export function parseToolcallActions(
  toolCalls: ToolCall[] | null | undefined,
  opts: { formatErrorTemplate: string; templateKwargs?: Record<string, unknown> },
): Action[] {
  const { formatErrorTemplate, templateKwargs = {} } = opts;
  if (!toolCalls || toolCalls.length === 0) {
    throw new FormatError({
      role: "user",
      content: nunjucks.renderString(formatErrorTemplate, {
        error: "No tool calls found in the response. Every response MUST include at least one tool call.",
        actions: [],
        has_tool_calls: false,
        ...templateKwargs,
      }),
      extra: { interrupt_type: "FormatError" },
    });
  }
  const actions: Action[] = [];
  for (const tc of toolCalls) {
    let errorMsg = "";
    let args: Record<string, unknown> = {};
    try {
      args = JSON.parse(tc.function.arguments);
    } catch (e) {
      errorMsg = `Error parsing tool call arguments: ${e}.`;
    }
    if (tc.function.name !== "bash") errorMsg += `Unknown tool '${tc.function.name}'.`;
    if (typeof args !== "object" || args === null || !("command" in args)) {
      errorMsg += "Missing 'command' argument in bash tool call.";
    }
    if (errorMsg) {
      throw new FormatError({
        role: "user",
        content: nunjucks.renderString(formatErrorTemplate, {
          actions: [],
          error: errorMsg.trim(),
          has_tool_calls: true,
          ...templateKwargs,
        }),
        extra: { interrupt_type: "FormatError" },
      });
    }
    actions.push({ command: args.command as string, tool_call_id: tc.id });
  }
  return actions;
}

/** Format execution outputs into tool result messages. */
export function formatToolcallObservationMessages(opts: {
  actions: Action[];
  outputs: EnvOutput[];
  observationTemplate: string;
  templateVars?: Record<string, unknown>;
  multimodalRegex?: string;
}): Message[] {
  const { actions, outputs, observationTemplate, templateVars = {}, multimodalRegex = "" } = opts;
  const notExecuted: EnvOutput = { output: "", returncode: -1, exception_info: "action was not executed" };
  const paddedOutputs = [...outputs, ...Array(Math.max(0, actions.length - outputs.length)).fill(notExecuted)];
  const results: Message[] = [];
  for (let i = 0; i < actions.length; i++) {
    const action = actions[i];
    const output = paddedOutputs[i];
    const content = nunjucks.renderString(observationTemplate, { output, ...templateVars });
    const msg: Message = {
      role: "",
      content,
      extra: {
        raw_output: output.output,
        returncode: output.returncode,
        timestamp: Date.now() / 1000,
        exception_info: output.exception_info,
        ...(output.extra || {}),
      },
    };
    if ("tool_call_id" in action && action.tool_call_id) {
      msg.tool_call_id = action.tool_call_id;
      msg.role = "tool";
    } else {
      msg.role = "user"; // human issued commands
    }
    results.push(multimodalRegex ? expandMultimodalContent(msg, multimodalRegex) : msg);
  }
  return results;
}


